src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,809 @@
+   * @param _tokenIds             Array of Tokens ID that are bought
+   * @param _tokensBoughtAmounts  Amount of Tokens id bought for each corresponding Token id in _tokenIds
+   * @param _maxCurrency          Total maximum amount of currency tokens to spend for all Token ids
+   * @param _deadline             Timestamp after which this transaction will be reverted
+   * @param _recipient            The address that receives output Tokens and refund
+   * @return currencySold How much currency was actually sold.
+   */
+  function _currencyToToken(
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokensBoughtAmounts,
+    uint256 _maxCurrency,
+    uint256 _deadline,
+    address _recipient)
+    internal nonReentrant() returns (uint256[] memory currencySold)
+  {
+    // Input validation
+    require(_deadline >= block.timestamp, "NiftyswapExchange20#_currencyToToken: DEADLINE_EXCEEDED");
+
+    // Number of Token IDs to deposit
+    uint256 nTokens = _tokenIds.length;
+    uint256 totalRefundCurrency = _maxCurrency;
+
+    // Initialize variables
+    currencySold = new uint256[](nTokens); // Amount of currency tokens sold per ID
+    uint256[] memory tokenReserves = new uint256[](nTokens);  // Amount of tokens in reserve for each Token id
+
+    // Get token reserves
+    tokenReserves = _getTokenReserves(_tokenIds);
+
+    // Assumes the currency Tokens are already received by contract, but not
+    // the Tokens Ids
+
+    // Remove liquidity for each Token ID in _tokenIds
+    for (uint256 i = 0; i < nTokens; i++) {
+      // Store current id and amount from argument arrays
+      uint256 idBought = _tokenIds[i];
+      uint256 amountBought = _tokensBoughtAmounts[i];
+      uint256 tokenReserve = tokenReserves[i];
+
+      require(amountBought > 0, "NiftyswapExchange20#_currencyToToken: NULL_TOKENS_BOUGHT");
+
+      // Load currency token and Token _id reserves
+      uint256 currencyReserve = currencyReserves[idBought];
+
+      // Get amount of currency tokens to send for purchase
+      // Neither reserves amount have been changed so far in this transaction, so
+      // no adjustment to the inputs is needed
+      uint256 currencyAmount = getBuyPrice(amountBought, currencyReserve, tokenReserve);
+
+      // Calculate currency token amount to refund (if any) where whatever is not used will be returned
+      // Will throw if total cost exceeds _maxCurrency
+      totalRefundCurrency = totalRefundCurrency.sub(currencyAmount);
+
+      // Append Token id, Token id amount and currency token amount to tracking arrays
+      currencySold[i] = currencyAmount;
+
+      // Update individual currency reseve amount
+      currencyReserves[idBought] = currencyReserve.add(currencyAmount);
+    }
+
+    // Refund currency token if any
+    if (totalRefundCurrency > 0) {
+      require(currency.transfer(_recipient, totalRefundCurrency), "NiftyswapExchange20#_currencyToToken: TRANSFER_FAILED");
```

Comment:
Some ERC20 tokens don't return true on `transfer` but some do, for maximum compatibility with tokens that don't fully comply with the standard we should use something like the Uniswap TransferHelper.

https://github.com/Uniswap/uniswap-lib/blob/master/contracts/libraries/TransferHelper.sol

Reply:
This applies to all the ERC20 transfers